### PR TITLE
Implemented Admin page MVP

### PR DESCRIPTION
**Pull request by DanielMcCrystal**
I know it looks like a ton but most of it is test cases which ended up being pretty tedious. Let me know if I need to break it up into smaller parts.

### CODE REVIEW

File: src/main/java/codeu/controller/AdminServlet.java [hunk 1]
```diff
@@ -1,50 +1,149 @@
...
+    int totalWordCount = 0; // total number of words sent by all users
+
+    for (Conversation c: allConversations) {
+
+      messagesInConversation = messageStore.getMessagesInConversation(c.getId());
+      totalMessageCount += messagesInConversation.size();
+
+      for (Message m: messagesInConversation) {
+
+        int tmp = userMessageCounts.getOrDefault(m.getAuthorId(), 0);
```
Comment (kyrabusser) [inline src/main/java/codeu/controller/AdminServlet.java]: `previousMessageCount` or `authorMessageCount` would be a more descriptive name than `tmp`.

File: src/main/java/codeu/controller/AdminServlet.java [hunk 2]
```diff
@@ -1,50 +1,149 @@
...
+      request.getRequestDispatcher("/WEB-INF/view/admin.jsp").forward(request, response);
+    } else {
+      response.sendError(HttpServletResponse.SC_NOT_FOUND);
     }
+
+  }
+
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response)
```
Comment (kyrabusser) [inline src/main/java/codeu/controller/AdminServlet.java]: If this servlet doesn't have a POST implementation yet, then no need to add this method.

File: src/main/java/codeu/controller/AdminServlet.java [hunk 3]
```diff
@@ -1,50 +1,149 @@
...
+    request.setAttribute("numberOfUsers", allUsers.size());
+    request.setAttribute("newestUser", newestUser);
+    request.setAttribute("mostActiveUser", mostActiveUser);
 
-    @Override
-    public void doPost(HttpServletRequest request, HttpServletResponse response)
-        throws IOException, ServletException {
+    request.setAttribute("numberOfConversations", allConversations.size());
+    request.setAttribute("numberOfMessages", totalMessageCount);
+    request.setAttribute("avgMessagesPerConvo", (int) ((double) totalMessageCount / allConversations.size()));
```
Comment (kyrabusser) [inline src/main/java/codeu/controller/AdminServlet.java]: Is this the behavior we want if the average isn't a whole number?

File: src/main/java/codeu/controller/AdminServlet.java [hunk 4]
```diff
@@ -1,50 +1,149 @@
...
+
+    // Users
+    List<User> allUsers = userStore.getAllUsers();
+
+    String newestUser = "N/A";
+    String mostActiveUser = "N/A";
+
+    // get newest user
+    if (allUsers.size() > 0) {
+      newestUser = allUsers.get(allUsers.size() - 1).getName();
```
Comment (kyrabusser) [inline src/main/java/codeu/controller/AdminServlet.java]: The User object has a creationTime field: https://github.com/Sereezy/CodeU-Spring-2018/blob/master/src/main/java/codeu/model/data/User.java#L58

Is getting the last user in the list guaranteed to always agree with those time stamps? I'm especially curious if we restart the server on AppEngine and load the users from the datastore. If we want this to be guaranteed, we should document it in the UserStore method.

Since this PR is pretty big already, if you decide this needs more work you could consider either removing the newest user field for now, or else adding a TODO comment that says we should investigate this further.

File: src/test/java/codeu/controller/AdminServletTest.java [hunk 5]
```diff
@@ -16,49 +18,454 @@
...
+    Mockito.verify(mockRequest).setAttribute("newestUser", "N/A");
+  }
+
+  @Test
+  public void testNewestUser_MultipleUsers() throws IOException, ServletException {
+    Mockito.when(mockSession.getAttribute("user")).thenReturn("admin");
+
+    User u1 = new User(UUID.randomUUID(), "user1", "hashed_pw", Instant.now());
+    User u2 = new User(UUID.randomUUID(), "user2", "hashed_pw", Instant.now());
+    User u3 = new User(UUID.randomUUID(), "user3", "hashed_pw", Instant.now());
```
Comment (kyrabusser) [inline src/test/java/codeu/controller/AdminServletTest.java]: See my comment in the code under test above--these users all have (approximately) the same creation timestamp. What happens if you manually give them different values? (`Instant.ofEpochSecond()` might be useful here.)

File: src/test/java/codeu/controller/AdminServletTest.java [hunk 6]
```diff
@@ -16,49 +18,454 @@
...
+    // Add mock messages
+    List<Message> mockMessagesInConversation = new ArrayList<Message>();
+    String[][] messages = {
+      {"a"},
+      {"a", "b"},
+      {"a", "b", "c"}
+    };
+    User[] users = {u1, u2, u3};
+
+    for (int u = 0; u < messages.length; u++) {
```
Comment (kyrabusser) [inline src/test/java/codeu/controller/AdminServletTest.java]: Hm...since it's important for the test that user3 have the most messages, it might be clearer if we add the conversations in a less nested way. It's probably also fine for this test if all of the message contents are the same, so that might make it a little simpler to set up the test data.

**Review by kyrabusser — CHANGES_REQUESTED** [review]
Wow, this is awesome! Can you include a screenshot of what the page looks like after generating some test data?

I LOVE your test coverage, especially that you remembered to check all the edge cases where the conversations or messages are empty.

In the future, if you'd like to keep the PR size smaller and simpler to review, you could split the changes out by feature--so separate PRs for number of users, newest user, most active user, conversations, etc. There's a little bit of an art to deciding which things make sense to combine and what the right size is, but smaller is generally better. :)

**Comment by DanielMcCrystal** [comment]
screen shot 2018-05-26 at 11 28 47 am
screen shot 2018-05-26 at 11 28 55 am
screen shot 2018-05-26 at 11 29 31 am
Here are some screenshots of the test data! I think I covered most of the issues you brought up in the new commits.

File: src/main/java/codeu/controller/AdminServlet.java [hunk 7]
```diff
@@ -1,50 +1,218 @@
...
+      request.getRequestDispatcher("/WEB-INF/view/admin.jsp").forward(request, response);
+    } else {
+      response.sendError(HttpServletResponse.SC_NOT_FOUND);
     }
+
+  }
+
+  /*
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response)
```
Comment (kyrabusser) [inline src/main/java/codeu/controller/AdminServlet.java]: Since this isn't being used, you should delete it, not just comment it out: https://blog.kentcdodds.com/please-don-t-commit-commented-out-code-53d0b5b26d5f

File: src/main/java/codeu/controller/AdminServlet.java [hunk 8]
```diff
@@ -1,50 +1,218 @@
...
+    request.setAttribute("mostActiveUser", mostActiveUser);
+
+    request.setAttribute("numberOfConversations", allConversations.size());
+    request.setAttribute("numberOfMessages", totalMessageCount);
+    request.setAttribute("avgMessagesPerConvo", avgMessagesPerConvo);
+    request.setAttribute("avgWordsPerMessage", avgWordsPerMessage);
+
+  }
+
+  void addTestData() {
```
Comment (kyrabusser) [inline src/main/java/codeu/controller/AdminServlet.java]: Wow, this is awesome! (I was just thinking of adding some users manually through the UI...) It's so awesome, in fact, that I think it deserves its own PR, tests, and maybe a separate class. It would be nice to make this behavior available to admin users, maybe with some customization options (how many items to add, for example).

In the meantime, can you remove this from the current PR? The commit itself is saved if you want to reference it later: https://github.com/Sereezy/CodeU-Spring-2018/commit/d942efc47805f81bd7d2ea7ee4078371d015eee0

File: src/main/java/codeu/controller/AdminServlet.java [hunk 9]
```diff
@@ -1,50 +1,218 @@
...
+    int numberOfUsers = ThreadLocalRandom.current().nextInt(10, 101);
+    int numberOfConversations = ThreadLocalRandom.current().nextInt(5, 21);
+    int numberOfMessages = ThreadLocalRandom.current().nextInt(100, 1001);
+
+    User newUser;
+    for (int i = 0; i < numberOfUsers; i++) {
+      newUser = new User(UUID.randomUUID(), "user"+i, "password",
+        Instant.ofEpochSecond(ThreadLocalRandom.current().nextInt(1, 10001)));
```
Comment (kyrabusser) [inline src/main/java/codeu/controller/AdminServlet.java]: Micronit: The Google style guide indents four spaces on line continuations like this, so it's clear that it's really just part of the line above it. Up to you if you'd like to change it (or use https://github.com/google/google-java-format).

**Review by kyrabusser — COMMENTED** [review]
Nice work! I like the descriptive, incremental commits too. 👍

**Review by kyrabusser — APPROVED** [review]
:+1:
